### ISSUE TITLE
Proposal: support wildcards and destination names  in `docker cp`

### ISSUE THREAD

**Issue by SvenDowideit** [docker/docker#7710]
Make `docker cp` work more like `scp`

`Usage: docker cp CONTAINER:PATH HOSTPATH`

## Proposed

1 Allow the `PATH` to contain standard filesystem wildcards, and then expand them on the container side.
2 Allow the `HOSTPATH` to be the resulting filename _if_ the daemon sends it a single file. 

## right now

`HOSTPATH` is a dir, which will even be created for you - rather than the name of the resulting file, and wildcards are not expanded.

```
tmp $ mkdir sven
tmp $ docker cp sven:/etc/* sven
2014/08/25 14:22:37 Error response from daemon: Could not find the file /etc/* in container sven
tmp $ docker cp sven:/etc/passwd sven/banana
tmp $ ls sven/
banana  passwd
tmp $ more sven/*

*** sven/banana: directory ***

tmp $ cd sven
sven $ docker cp sven:/etc/* .
2014/08/25 14:22:23 Error response from daemon: Could not find the file /etc/* in container sven
```

**Comment by ferrouswheel** [comment]
@duglin I am unable to find anything about `docker cp` supporting wildcards, and am unable to get wildcards working with docker 1.5.0. Can you please point towards me to somewhere with more info?

**Comment by gpetz** [comment]
@duglin 
``
docker cp 87ff9d25b53d:/var/log/apache2/access.log .
``
copies access.log to .

``
docker cp 87ff9d25b53d:/var/log/apache2/*.log .
``
returns: ATA[0000] Error response from daemon: Could not find the file /var/log/apache2/*.log in container 87ff9d25b53d

(Boot2Docker v1.6.2)

**Comment by duglin** [comment]
You're correct - I think was mixing up 'docker cp' with "COPY" from Dockerfiles.
However #12858 should do it.

**Comment by mmarzec** [comment]
+1 for this feature

**Comment by xmhubj** [comment]
looking forward to be implemented.

**Comment by francoishill** [comment]
+1

**Comment by deiga** [comment]
What is the current status of this issue? Following all the PR links makes it seem like nothing has happened to this, but I can't be sure.
We're talking about a 3 years old request here.

**Comment by thaJeztah** [comment]
> What is the current status of this issue?

It's still open, I don't think someone looked into this, but contributions welcome. Implementation may be non-trivial though (i.e.; correctly implementing wildcard matching has proven to be tricky)

**Comment by javadba** [comment]
Can an entire directory recursively be copied?  Otherwise .. the obvious question.. how can anything be bulk copied?

**Comment by thaJeztah** [comment]
> Can an entire directory recursively be copied? Otherwise .. the obvious question.. how can anything be bulk copied?

recursive copy is supported;

prepare a container having some files;

```
mkdir repro-7710 && cd repro-7710
docker run --name mycontainer busybox sh -c 'mkdir -p /foo/bar/baz; touch /foo/file /foo/bar/file /foo/bar/baz/file;'
```

copy the _contents_ of `foo` recursively;

```
mkdir output
docker cp mycontainer:/foo/. ./output
tree output
output
├── bar
│   ├── baz
│   │   └── file
│   └── file
└── file

2 directories, 3 files
```

copy `foo`, and its _contents_ recursively;

```
mkdir output2
docker cp mycontainer:/foo ./output2/
tree output2
output2
└── foo
    ├── bar
    │   ├── baz
    │   │   └── file
    │   └── file
    └── file

3 directories, 3 files
```

**Comment by courtarro** [comment]
A workaround for these limitations, at least with `aufs`, is to find the mount point of the container's path and copy files from there:

 - Find the ID of the `aufs` mount point for this container: `sudo cat /var/lib/docker/image/aufs/layerdb/mounts/(container_id)/mount-id`
 - Navigate to the path (as root). This requires that the container be running so that it's mounted: `cd /var/lib/docker/aufs/mnt/(mount_id)`
 - Copy to/from this path

**Comment by michaelarnauts** [comment]
Until this is implemented, I use the following workaround by tarring it in the container, and extracting it on the host.

```shell
docker exec *container* sh -c "tar -cf - /var/log/queue.*.log" | tar --strip-components=2 -xf - -C output/
```

**Comment by Aypahyo** [comment]
I have a immilar usecase for this.
I have output files that I need to copy.

the names in the container are:
./src/IntegrationTests/TestResults/TrxResults_2019-05-28_15-06-19-518.xml
./src/IntegrationTests/TestResults/TrxResults_2019-05-31_08-08-42-488.xml
./src/UnitTests/TestResults/TrxResults_2019-05-28_15-06-21-358.xml
./src/UnitTests/TestResults/TrxResults_2019-05-31_08-08-45-114.xml
./src/UnitTests/coverage.opencover.xml
./src/.sonarqube/out/.sonar/report-task.txt

My container is not running so an exec hotfix can not be done.

These commands would find the file in the container when I use an interactive terminal.
find . -wholename "**/TestResults/*.xml"
find . -wholename "**/coverage.opencover.xml"
find . -name "report-task.txt"

Following the pattern:
docker cp [OPTIONS] CONTAINER:SRC_PATH DEST_PATH

I would like to write
docker cp -wholename "**/TestResults/*.xml"  CONTAINER:src .
docker cp -wholename "**/coverage.opencover.xml"  CONTAINER:src .
docker cp -name "report-task.txt"  CONTAINER:src .

With a limitation that SRC_PATH and DEST_PATH can only be folder names.
The subfolder structure would be copied from the detected path.

The expected copy result would be 
DEST_PATH/IntegrationTests/TestResults/TrxResults_2019-05-28_15-06-19-518.xml
DEST_PATH/IntegrationTests/TestResults/TrxResults_2019-05-31_08-08-42-488.xml
DEST_PATH/UnitTests/TestResults/TrxResults_2019-05-28_15-06-21-358.xml
DEST_PATH/UnitTests/TestResults/TrxResults_2019-05-31_08-08-45-114.xml
DEST_PATH/UnitTests/coverage.opencover.xml
DEST_PATH/.sonarqube/out/.sonar/report-task.txt

**Comment by deiga** [comment]
@Aypahyo You can circumvent the "not running" part by commiting the container to an image and running the exec workaround on the new image.

**Comment by jpereira** [comment]
+1 for this feature

**Comment by vinaygunnam** [comment]
One other alternative to achieve what is being asked from this enhancement is to use a `dockerignore` and `docker cp` together. I would think that this is far more powerful than equipping `docker cp` with wildcard support.

As an example, @Aypahyo's [use case](https://github.com/moby/moby/issues/7710#issuecomment-497626937) can be solved using this inverse dockerignore file to whitelist the files of interest.

File: `Dockerfile.copy.dockerignore`
```
# ignore everything
*

# except these artifacts of interest
!**/TestResults/*.xml
!**/coverage.opencover.xml
!**/report-task.txt
```

Essentially, we can docker to create a context (via daemon sock bind + docker_cli) from inside the final container to suit our needs and copy it to the host via `docker cp`.
File: `Dockerfile.copy`
```
ARG RESULT_IMAGE
FROM ${RESULT_IMAGE}
WORKDIR /temporary

COPY . .
```

Finally,
```
docker cp <COPY_IMAGE_CONTAINER>:/temporary/src <path-mapped-to-the-host-target>
```

This allows to address any complex "copy the result" scenarios using Docker's ability to tailor a context as needed.

I'm no expert with Docker, so I'm looking for any feedback folks might have. Does this approach fall into one of these buckets?

1. This is a bad idea and we should not use this because _____ (ex: perf, security, etc.).
2. This does not work for all use cases such as ____.
3. This is too complex and/or an overkill to solve the problem at hand.

**Comment by signalogic** [comment]
Evidently, after 7.x years, as of 2022, the docker cp command still does not support wildcards. Is it possible to at least give a better error message such as ""docker cp" requires exactly 2 arguments" ? That would at least save time vs hunting for status of wildcard support on various forums

**Comment by kevinkopf** [comment]
> Evidently, after 7.x years, as of 2022, the docker cp command still does not support wildcards. Is it possible to at least give a better error message than ""docker cp" requires exactly 2 arguments" ? That would at least save time vs hunting for status of wildcard support on various forums

One more year
And we're still there
Whilst I'm getting older
Getting gray hair

**Comment by proof-of-reality** [comment]
Wow, this is wild !

**Comment by debilin** [comment]
+1 :pray:

**Comment by iamjjanga-ouo** [comment]
I'm also stucked... is there anybody in 2025? ping..

**Comment by cektom** [comment]
+1 🙏

**Comment by StiliyanKushev** [comment]
+1
